readme.md
# Backbone fetch cache

[![Build Status](https://travis-ci.org/mrappleton/backbone-fetch-cache.png?branch=master)](https://travis-ci.org/mrappleton/backbone-fetch-cache)

A Backbone plugin to cache calls to `Backbone.Model.prototype.fetch` and
`Backbone.Collection.prototype.fetch` in memory and `localStorage`.

## How it works
This plugin intercepts calls to `fetch` and stores the results in a cache object (`Backbone.fetchCache._cache`). If fetch is called with `{ cache: true }` in the options and the URL has already been cached the AJAX call will be skipped.

The local cache is persisted in `localStorage` if available for faster initial page loads.

The `prefill` option allows for models and collections to be filled with cache data just until the `fetch` operations are complete -- a nice way to make the app feel snappy on slower connections.

## What's wrong with browser caching for AJAX responses?
Nothing. This plugin is primarily for working with an API where you don't have control over response cache headers.

## Usage
Add the script to the page after backbone.js has been included:

```html
<script src="/path/to/backbone.js"></script>
<script src="/path/to/backbone.fetch-cache.js"></script>
```

or if you're using AMD, require the script as a module:

```js
require(['path/to/backbone.fetch-cache.js']);
```

Note that the AMD module depends on `underscore` and `backbone` modules being defined as it lists them as dependencies. If you don't have these mapped, you can do it by adding the following to your require config:

```js
requirejs.config({
  paths: {
    backbone: 'actual/path/to/backbone.js',
    underscore: 'actual/path/to/underscore.js'
  }
});
```

A note on [Zepto.js](http://zeptojs.com/). This plugin uses `jQuery.Deferred`
which is not included in Zepto. You'll need to add a third party
implementation of `jQuery.Deferred`, e.g. [Standalone-Deferred](https://github.com/Mumakil/Standalone-Deferred)

## Options
### `cache`
Calls to `modelInstance.fetch` or `collectionInstance.fetch` will be fulfilled from the cache (if possible) when `cache: true` is set in the options hash:

```js
myModel.fetch({ cache: true });
myCollection.fetch({ cache: true });
```

<hr />

### `prefill`
This option allows the model/collection to be populated from the cache immediately and then be updated once the call to `fetch` has completed. The initial cache hit calls the `prefillSuccess` callback and then the AJAX success/error callbacks are called as normal when the request is complete. This allows the page to render something immediately and then update it after the request completes. (Note: the `prefillSuccess` callback _will not fire_ if the data is not found in the cache.)

```js
myModel.fetch({
  prefill: true,
  prefillSuccess: someCallback, // Fires when the cache hit happens
  success: anotherCallback // Fires after the AJAX call
});

myCollection.fetch({
  prefill: true,
  prefillSuccess: someCallback, // Fires when the cache hit happens
  success: anotherCallback // Fires after the AJAX call
});
```

This option can be used with the promises interface like so (note: the `progress` event _will not fire_ if the data is not found in the cache.):

```js
var modelPromise = myModel.fetch({ prefill: true });
modelPromise.progress(someCallback); // Fires when the cache hit happens
modelPromise.done(anotherCallback); // Fires after the AJAX call

var collectionPromise = myModel.fetch({ prefill: true });
collectionPromise.progress(someCallback); // Fires when the cache hit happens
collectionPromise.done(anotherCallback); // Fires after the AJAX call
```

<hr />

### `expires`

Cache vales expire after 5 minutes by default. You can adjust this by passing
`expires: <seconds>` to the fetch call. Set to `false` to never expire:

```js
myModel.fetch({ cache: true, expires: 60000 });
myCollection.fetch({ cache: true, expires: 60000 });

// These will never expire
myModel.fetch({ cache: true, expires: false });
myCollection.fetch({ cache: true, expires: false });
```

<hr />

### `localStorage`
By default the cache is persisted in localStorage (if available). Set `Backbone.fetchCache.localStorage = false` to disable this:

```js
Backbone.fetchCache.localStorage = false;
```

### Custom cache keys

By default the cache key is generated from the model's `url` property and the requests params:

```js
'/model/1?some=param'
```

This can be overridden with custom logic if required:

```js
// Instance is a Backbone.Model or Backbone.Collection, options are passed
// through form the fetch call
Backbone.fetchCache.getCacheKey = function(instance, options) {
  return instance.constructor.name + ':' + instance.get('id');
  // => UserModel:1
};
```

### `priorityFn`
When setting items in localStorage, the browser may throw a ```QUOTA_EXCEEDED_ERR```, meaning the store is full. Backbone.fetchCache tries to work around this problem by deleting what it considers the most stale item to make space for the new data. The staleness of data is determined by the sorting function ```priorityFn````, which by default returns the oldest item.

The default is:
```
Backbone.fetchCache.priorityFn = function(a, b) {
  if (!a || !a.expires || !b || !b.expires) {
    return a;
  }

  return a.expires - b.expires;
};
```

You can override this function with your own logic (in this case, returning the most recent item):
```
Backbone.fetchCache.priorityFn = function(a, b) {
  return b.expires - a.expires;
};
```

## Events

The `sync` event is triggered on a cache hit. A `cachesync` event is also triggered on Models and Collections, but only when a cache hit happens. This can be used if you need to differentiate between a server backed `sync` event and a cache backed event.

## Automatic cache invalidation
The cache item for a particular call will be cleared when a `create`, `update`, `patch` or `delete` call is made to the server. The plugin tries to be intelligent about this by clearing a model's collection cache if the model has a `.collection property`.

To achieve this, the plugin overrides `Backbone.Model.protoype.sync` and then calls the original method. If you are planning to override sync on a particular model then you should keep this in mind and make sure that you do it before the plugin runs. Overriding Backbone.sync directly should work fine.

## Manual cache invalidation
Sometimes you just need to clear a cached item manually. This can be called safely from anywhere in your application.

```js
Backbone.fetchCache.clearItem(myModel.url);
```

## Tests
You can run the tests by cloning the repo, installing the dependencies and
running `grunt jasmine`:

```
$ npm install
$ grunt jasmine
```

The default grunt task runs tests and lints the code.

```
$ grunt
```


CHANGELOG
Changelog

- v1.1.2: Bugfix - Collection fetch now triggers both the collection's parse method, and the model's parse method - thanks to [@BananaNeil](https://github.com/BananaNeil)
- v1.1.1: Bugfix - respect jQuery.ajax's async: true option and resolve from the cache synchronously if required - thanks to [@matthewkremer](https://github.com/matthewkremer)
- v1.1.0: Change - Don't rely on global jQuery. We now use Backbone.root.jQuery or an AMD defined `jquery` module - thanks to [@brett-shwom](https://github.com/brett-shwom)
- v1.0.2: Bugfix - adds better check for localStorage support - thanks to [@jfmoy](https://github.com/jfmoy)
- v1.0.1: Bugfix - updates fetch success callback signature to match Backbone's - thanks to [@ydaniv](https://github.com/ydaniv)
- v1.0.0: Perform all cache fulfilment asynchronously. Only the promise, not the full $.Deferred object is returned from fetch functions.

- v0.1.11: Account for custom URLs to be passed in fetch method options - thanks to [@dincho](https://github.com/dincho)
- v0.1.10: Add getCacheKey function to allow custom cache keys - thanks to [@ebertti](https://github.com/ebertti)
- v0.1.9: Fixes a bug where clearing a cache item didn't get persisted to localStorage - thanks to [@mzafer](https://github.com/mzafer)
- v0.1.8: Trigger `sync` and `cachesync` events - thanks to [@victorquinn](https://github.com/victorquinn).
- v0.1.7: Ensures that `parse` is called for collections - thanks to [@fernandopg](https://github.com/fernandopg).
- v0.1.6: Update to call the correct `add` and `reset` methods for collections - thanks to [@mzafer](https://github.com/mzafer).
- v0.1.5: Attempt to clear cache keys on PUT/POST/DELETE
- v0.1.4: Bug fix for older Webkits which throw an error trying to JSON.parse(null)
- v0.1.3: Auto expiration of old cache items if `locaStorage` gets full - thanks to [@inf0rmer](https://github.com/inf0rmer).
- v0.1.2: Add AMD support.
- v0.1.1: Add `prefetch` option.


spec/model.fetch.spec.js
describe('Backbone.Model', function() {
  beforeEach(function() {
    this.model = new Backbone.Model();
    this.model.url = '/model-cache-test';

    // Mock xhr resposes
    this.server = sinon.fakeServer.create();
    this.response = { sausages: 'bacon' };
    this.server.respondWith('GET', this.model.url, [
      200,
      { "Content-Type": "application/json" },
      JSON.stringify(this.response)
    ]);
  });

  afterEach(function() {
    Backbone.Model.attributeCache = {};
    localStorage.clear('modelCache');
    this.server.restore();
  });

  describe('.setCache', function() {
    it('noops if the instance does not have a url', function() {
      this.model.url = null;
      Backbone.Model.setCache(this.model);
      expect(Backbone.Model.attributeCache[this.model.url]).toBeUndefined();
    });

    it('keys cache items by URL',function() {
      Backbone.Model.setCache(this.model);
      expect(Backbone.Model.attributeCache[this.model.url].value).toEqual(this.model.toJSON());
    });

    it('calls setLocalStorage', function() {
      spyOn(Backbone.Model, 'setLocalStorage');
      Backbone.Model.setCache(this.model);
      expect(Backbone.Model.setLocalStorage).toHaveBeenCalled();
    });

    describe('cache expiry', function() {
      beforeEach(function() {
        this.clock = sinon.useFakeTimers();
      });

      afterEach(function() {
        this.clock.restore();
      });

      it('sets default expiry times for cache keys', function() {
        Backbone.Model.setCache(this.model, { cache: true });
        expect(Backbone.Model.attributeCache[this.model.url].expires)
          .toEqual((new Date()).getTime() + (5* 60 * 1000));
      });

      it('sets expiry times for cache keys', function() {
        var opts = { cache: true, expires: 1000 };
        Backbone.Model.setCache(this.model, opts);
        expect(Backbone.Model.attributeCache[this.model.url].expires)
          .toEqual((new Date()).getTime() + (opts.expires * 1000));
      });

      it('is not set if expires: false is set', function() {
        var opts = { cache: true, expires: false };
        Backbone.Model.setCache(this.model, opts);
        expect(Backbone.Model.attributeCache[this.model.url].expires)
          .toEqual(false);
      });
    });
  });

  describe('.setLocalStorage', function() {
    it('puts the cache into localStorage', function() {
      var cache = Backbone.Model.attributeCache = {
        '/url1': { expires: false, value: { bacon: 'sandwich' } },
        '/url2': { expires: false, value: { egg: 'roll' } }
      };
      Backbone.Model.setLocalStorage();
      expect(localStorage.getItem('modelCache')).toEqual(JSON.stringify(cache));
    });

    it('does not put the cache into localStorage if localStorageCache = false', function() {
      var cache = Backbone.Model.attributeCache = {
        '/url1': { expires: false, value: { bacon: 'sandwich' } },
        '/url2': { expires: false, value: { egg: 'roll' } }
      };
      Backbone.Model.localStorageCache = false;
      Backbone.Model.setLocalStorage();

      expect(localStorage.getItem('modelCache')).toEqual(null);

      Backbone.Model.localStorageCache = true; // restore
    });
  });


  describe('.getLocalStorage', function() {
    it('primes the cache from localStorage', function() {
      var cache = {
        '/url1': { expires: false, value: { bacon: 'sandwich' } },
        '/url2': { expires: false, value: { egg: 'roll' } }
      };
      localStorage.setItem('modelCache', JSON.stringify(cache));
      Backbone.Model.getLocalStorage();
      expect(Backbone.Model.attributeCache).toEqual(cache);
    });
  });

  describe('.prototype.fetch', function() {
    it('saves returned attributes to the attributeCache', function() {
      this.model.fetch();
      this.server.respond();
      expect(Backbone.Model.attributeCache[this.model.url].value).toEqual(this.model.toJSON());
    });

    it('passes the instance and options through to setCache', function() {
      var opts = { banana: 'bread' };
      spyOn(Backbone.Model, 'setCache');

      this.model.fetch(opts);
      this.server.respond();

      expect(Backbone.Model.setCache.calls[0].args[0]).toEqual(this.model);
      expect(Backbone.Model.setCache.calls[0].args[1]).toEqual(opts);
    });

    it('returns data from the cache if cache: true is set', function() {
      var cacheData = { cheese: 'pickle' };
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };
      this.model.fetch({ cache: true });
      expect(this.model.toJSON()).toEqual(cacheData);
    });

    it('does not return cache data if cache: true is not set', function() {
      var cacheData = { cheese: 'pickle' };
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.model.fetch();
      this.server.respond();

      expect(this.model.toJSON()).not.toEqual(cacheData);
      expect(this.model.toJSON()).toEqual(this.response);
    });

    it('does not return cache data if the cache item is stale', function() {
      var cacheData = { cheese: 'pickle' };
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() - (5* 60 * 1000)
      };

      this.model.fetch();
      this.server.respond();

      expect(this.model.toJSON()).not.toEqual(cacheData);
      expect(this.model.toJSON()).toEqual(this.response);
    });

    it('returns cache data if the item has expires: false', function() {
      var cacheData = { cheese: 'pickle' };
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: false
      };

      this.model.fetch({ cache: true });

      expect(this.model.toJSON()).toEqual(cacheData);
    });

    it('calls success callback on a cache hit', function() {
      var success = jasmine.createSpy('success'),
          cacheData = { cheese: 'pickle' };
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.model.fetch({ cache: true, success: success });

      expect(success).toHaveBeenCalledWith(this.model);
    });

    it('returns a fulfilled promise on a cache hit', function() {
      var cacheData = { cheese: 'pickle' },
          promise;
      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };
      promise = this.model.fetch({ cache: true });

      expect(promise).toBeAPromise();
      expect(promise).toBeResolved();
    });

    it('returns an unfulfilled promise on a cache miss', function() {
      var promise = this.model.fetch();
      expect(promise).toBeAPromise();
      expect(promise).toBeUnresolved();
    });

    it('returns a promise with the correct context on a cache hit', function() {
      var cacheData = { cheese: 'pickle' },
          spy = jasmine.createSpy('success'),
          opts = { cache: true };

      Backbone.Model.attributeCache[this.model.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.model.fetch(opts).done(spy);

      expect(spy).toHaveBeenCalledWith(this.model);
    });
  });
});

spec/support/matchers.js
(function() {
  function isFunc (object) {
    return typeof object === 'function';
  }

  beforeEach(function() {
    this.addMatchers({
      toBeAPromise: function(expected) {
        return (
          isFunc(this.actual.state) &&
          isFunc(this.actual.always) &&
          isFunc(this.actual.then) &&
          isFunc(this.actual.promise) &&
          isFunc(this.actual.pipe) &&
          isFunc(this.actual.done) &&
          isFunc(this.actual.fail) &&
          isFunc(this.actual.progress)
        );
      },
      toBeResolved: function(expected)  {
        return this.actual.state() === 'resolved';
      }
    });
  });
})();


spec/collection.fetch.spec.js
describe('Backbone.Collection', function() {
  beforeEach(function() {
    this.collection = new Backbone.Collection();
    this.collection.url = '/collection-cache-test';

    // Mock xhr resposes
    this.server = sinon.fakeServer.create();
    this.response = [{ sausages: 'bacon' }, { rice: 'peas' }];
    this.server.respondWith('GET', this.collection.url, [
      200,
      { "Content-Type": "application/json" },
      JSON.stringify(this.response)
    ]);
  });

  afterEach(function() {
    Backbone.Collection.attributeCache = {};
    localStorage.clear('collectionCache');
    this.server.restore();
  });

  describe('.setCache', function() {
    it('noops if the instance does not have a url', function() {
      this.collection.url = null;
      Backbone.Collection.setCache(this.collection);
      expect(Backbone.Collection.attributeCache).toEqual({});
    });

    it('keys cache items by URL', function() {
      Backbone.Collection.setCache(this.collection);
      expect(Backbone.Collection.attributeCache[this.collection.url].value)
        .toEqual(this.collection.toJSON());
    });

    describe('cache expiry', function() {
      beforeEach(function() {
        this.clock = sinon.useFakeTimers();
      });

      afterEach(function() {
        this.clock.restore();
      });

      it('sets default expiry times for cache keys', function() {
        Backbone.Collection.setCache(this.collection, { cache: true });
        expect(Backbone.Collection.attributeCache[this.collection.url].expires)
          .toEqual((new Date()).getTime() + (5* 60 * 1000));
      });

      it('sets expiry times for cache keys', function() {
        var opts = { cache: true, expires: 1000 };
        Backbone.Collection.setCache(this.collection, opts);
        expect(Backbone.Collection.attributeCache[this.collection.url].expires)
          .toEqual((new Date()).getTime() + (opts.expires * 1000));
      });

      it('is not set if expires: false is set', function() {
        var opts = { cache: true, expires: false };
        Backbone.Collection.setCache(this.collection, opts);
        expect(Backbone.Collection.attributeCache[this.collection.url].expires)
          .toEqual(false);
      });

      it('calls setLocalStorage', function() {
        spyOn(Backbone.Collection, 'setLocalStorage');
        Backbone.Collection.setCache(this.collection);
        expect(Backbone.Collection.setLocalStorage).toHaveBeenCalled();
      });
    });
  });

  describe('.setLocalStorage', function() {
    it('puts the cache into localStorage', function() {
      var cache = Backbone.Collection.attributeCache = {
        '/url1': [{ expires: false, value: { bacon: 'sandwich' } }],
        '/url2': [{ expires: false, value: { egg: 'roll' } }]
      };
      Backbone.Collection.setLocalStorage();
      expect(localStorage.getItem('collectionCache')).toEqual(JSON.stringify(cache));
    });

    it('does not put the cache into localStorage if localStorageCache = false', function() {
      var cache = Backbone.Collection.attributeCache = {
        '/url1': { expires: false, value: { bacon: 'sandwich' } },
        '/url2': { expires: false, value: { egg: 'roll' } }
      };
      Backbone.Collection.localStorageCache = false;
      Backbone.Collection.setLocalStorage();

      expect(localStorage.getItem('collectionCache')).toEqual(null);

      Backbone.Collection.localStorageCache = true; // restore
    });
  });

  describe('.getLocalStorage', function() {
    it('primes the cache from localStorage', function() {
      var cache = {
        '/url1': [{ expires: false, value: { bacon: 'sandwich' } }],
        '/url2': [{ expires: false, value: { egg: 'roll' } }]
      };
      localStorage.setItem('collectionCache', JSON.stringify(cache));
      Backbone.Collection.getLocalStorage();
      expect(Backbone.Collection.attributeCache).toEqual(cache);
    });

    it('does not prime the cache from localStorage if localStorageCache = false', function() {
      var cache = Backbone.Collection.attributeCache = {
        '/url1': [{ expires: false, value: { bacon: 'sandwich' } }],
        '/url2': [{ expires: false, value: { egg: 'roll' } }]
      };
      localStorage.setItem('collectionCache', JSON.stringify(cache));

      Backbone.Collection.attributeCache = {};
      Backbone.Collection.localStorageCache = false;

      Backbone.Collection.getLocalStorage();

      expect(Backbone.Collection.attributeCache).toEqual({});

      Backbone.Collection.localStorageCache = true; // restore
    });
  });

  describe('.prototype.fetch', function() {
    it('saves returned attributes to the attributeCache', function() {
      this.collection.fetch();
      this.server.respond();
      expect(Backbone.Collection.attributeCache[this.collection.url].value)
        .toEqual(this.collection.toJSON());
    });

    it('passes the instance and options through to setCache', function() {
      var opts = { banana: 'bread' };
      spyOn(Backbone.Collection, 'setCache');

      this.collection.fetch(opts);
      this.server.respond();

      expect(Backbone.Collection.setCache.calls[0].args[0]).toEqual(this.collection);
      expect(Backbone.Collection.setCache.calls[0].args[1]).toEqual(opts);
    });

    it('returns data from the cache if cache: true is set', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }];
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };
      this.collection.fetch({ cache: true });
      expect(this.collection.toJSON()).toEqual(cacheData);
    });

    it('does not return cache data if cache: true is not set', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }];
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.collection.fetch();
      this.server.respond();

      expect(this.collection.toJSON()).not.toEqual(cacheData);
      expect(this.collection.toJSON()).toEqual(this.response);
    });

    it('does not return cache data if the cache item is stale', function() {
      var cacheData = { cheese: 'pickle' };
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() - (5* 60 * 1000)
      };

      this.collection.fetch();
      this.server.respond();

      expect(this.collection.toJSON()).not.toEqual(cacheData);
      expect(this.collection.toJSON()).toEqual(this.response);
    });

    it('returns cache data if the item has expires: false', function() {
      var cacheData = [{ cheese: 'pickle' }];
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: false
      };

      this.collection.fetch({ cache: true });

      expect(this.collection.toJSON()).toEqual(cacheData);
    });

    it('calls add according to options on a cache hit', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }],
          options = { cache: true, add: true };

      spyOn(this.collection, 'add');
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.collection.fetch(options);

      expect(this.collection.add).toHaveBeenCalledWith(cacheData, options);
    });

    it('calls reset according to options on a cache hit', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }],
          options = { cache: true, add: false };

      spyOn(this.collection, 'reset');
      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.collection.fetch(options);

      expect(this.collection.reset).toHaveBeenCalledWith(cacheData, options);
    });

    it('calls success callback on a cache hit', function() {
      var success = jasmine.createSpy('success'),
          cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }];

      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.collection.fetch({ cache: true, success: success });

      expect(success).toHaveBeenCalledWith(this.collection);
    });

    it('returns a fulfilled promise on a cache hit', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }],
          promise;

      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };
      promise = this.collection.fetch({ cache: true });

      expect(promise).toBeAPromise();
      expect(promise).toBeResolved();
    });

    it('returns an unfulfilled promise on a cache miss', function() {
      var promise = this.collection.fetch();
      expect(promise).toBeAPromise();
      expect(promise).toBeUnresolved();
    });

    it('returns a promise with the correct context on a cache hit', function() {
      var cacheData = [{ cheese: 'pickle' }, { salt: 'vinegar' }],
          spy = jasmine.createSpy('success');

      Backbone.Collection.attributeCache[this.collection.url] = {
        value: cacheData,
        expires: (new Date()).getTime() + (5* 60 * 1000)
      };

      this.collection.fetch({ cache: true }).done(spy);

      expect(spy).toHaveBeenCalledWith(this.collection);
    });

  });
});

backbone.fetch-cache.js
/*!
  backbone.fetch-cache v1.1.1
  by Andy Appleton - https://github.com/mrappleton/backbone-fetch-cache.git
 */

// AMD wrapper from https://github.com/umdjs/umd/blob/master/amdWebGlobal.js

(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module and set browser global
    define(['underscore', 'backbone', 'jquery'], function (_, Backbone, $) {
      return (root.Backbone = factory(_, Backbone, $));
    });
  } else {
    // Browser globals
    root.Backbone = factory(root._, root.Backbone, root.jQuery);
  }
}(this, function (_, Backbone, $) {

  // Setup
  var superMethods = {
    modelFetch: Backbone.Model.prototype.fetch,
    modelSync: Backbone.Model.prototype.sync,
    collectionFetch: Backbone.Collection.prototype.fetch
  },
  supportLocalStorage = (function() {
    var supported = typeof window.localStorage !== 'undefined';
    if (supported) {
      try {
        // impossible to write on some platforms when private browsing is on and
        // throws an exception = local storage not supported.
        localStorage.setItem("test_support", "test_support");
        localStorage.removeItem("test_support");
      } catch (e) {
        supported = false;
      }
    }
    return supported;
  })();

  Backbone.fetchCache = (Backbone.fetchCache || {});
  Backbone.fetchCache._cache = (Backbone.fetchCache._cache || {});

  Backbone.fetchCache.priorityFn = function(a, b) {
    if (!a || !a.expires || !b || !b.expires) {
      return a;
    }

    return a.expires - b.expires;
  };

  Backbone.fetchCache._prioritize = function() {
    var sorted = _.values(this._cache).sort(this.priorityFn);
    var index = _.indexOf(_.values(this._cache), sorted[0]);
    return _.keys(this._cache)[index];
  };

  Backbone.fetchCache._deleteCacheWithPriority = function() {
    Backbone.fetchCache._cache[this._prioritize()] = null;
    delete Backbone.fetchCache._cache[this._prioritize()];
    Backbone.fetchCache.setLocalStorage();
  };

  if (typeof Backbone.fetchCache.localStorage === 'undefined') {
    Backbone.fetchCache.localStorage = true;
  }

  // Shared methods
  function getCacheKey(instance, opts) {
    var url;

    if(opts && opts.url) {
      url = opts.url;
    } else {
      url = _.isFunction(instance.url) ? instance.url() : instance.url;
    }

    // Need url to use as cache key so return if we can't get it
    if(!url) { return; }

    if(opts && opts.data) {
      return url + "?" + $.param(opts.data);
    }
    return url;
  }

  function setCache(instance, opts, attrs) {
    opts = (opts || {});
    var key = Backbone.fetchCache.getCacheKey(instance, opts),
        expires = false;

    // Need url to use as cache key so return if we can't get it
    if (!key) { return; }

    // Never set the cache if user has explicitly said not to
    if (opts.cache === false) { return; }

    // Don't set the cache unless cache: true or prefill: true option is passed
    if (!(opts.cache || opts.prefill)) { return; }

    if (opts.expires !== false) {
      expires = (new Date()).getTime() + ((opts.expires || 5 * 60) * 1000);
    }

    Backbone.fetchCache._cache[key] = {
      expires: expires,
      value: attrs
    };

    Backbone.fetchCache.setLocalStorage();
  }

  function clearItem(key) {
    delete Backbone.fetchCache._cache[key];
    Backbone.fetchCache.setLocalStorage();
  }

  function setLocalStorage() {
    if (!supportLocalStorage || !Backbone.fetchCache.localStorage) { return; }
    try {
      localStorage.setItem('backboneCache', JSON.stringify(Backbone.fetchCache._cache));
    } catch (err) {
      var code = err.code || err.number || err.message;
      if (code === 22) {
        this._deleteCacheWithPriority();
      } else {
        throw(err);
      }
    }
  }

  function getLocalStorage() {
    if (!supportLocalStorage || !Backbone.fetchCache.localStorage) { return; }
    var json = localStorage.getItem('backboneCache') || '{}';
    Backbone.fetchCache._cache = JSON.parse(json);
  }

  function nextTick(fn) {
    return window.setTimeout(fn, 0);
  }

  // Instance methods
  Backbone.Model.prototype.fetch = function(opts) {
    opts = _.defaults(opts || {}, { parse: true });
    var key = Backbone.fetchCache.getCacheKey(this, opts),
        data = Backbone.fetchCache._cache[key],
        expired = false,
        attributes = false,
        deferred = new $.Deferred(),
        self = this;

    function setData() {
      self.set(attributes, opts);
      if (_.isFunction(opts.prefillSuccess)) { opts.prefillSuccess(self, attributes, opts); }

      // Trigger sync events
      self.trigger('cachesync', self, attributes, opts);
      self.trigger('sync', self, attributes, opts);

      // Notify progress if we're still waiting for an AJAX call to happen...
      if (opts.prefill) { deferred.notify(self); }
      // ...finish and return if we're not
      else {
        if (_.isFunction(opts.success)) { opts.success(self, attributes, opts); }
        deferred.resolve(self);
      }
    }

    if (data) {
      expired = data.expires;
      expired = expired && data.expires < (new Date()).getTime();
      attributes = data.value;
    }

    if (!expired && (opts.cache || opts.prefill) && attributes) {
      // Ensure that cache resolution adhers to async option, defaults to true.
      if (opts.async == null) { opts.async = true; }

      if (opts.async) {
        nextTick(setData);
      } else {
        setData();
      }

      if (!opts.prefill) {
        return deferred.promise();
      }
    }

    // Delegate to the actual fetch method and store the attributes in the cache
    superMethods.modelFetch.apply(this, arguments)
      // resolve the returned promise when the AJAX call completes
      .done( _.bind(deferred.resolve, this, this) )
      // Set the new data in the cache
      .done( _.bind(Backbone.fetchCache.setCache, null, this, opts) )
      // Reject the promise on fail
      .fail( _.bind(deferred.reject, this, this) );

    // return a promise which provides the same methods as a jqXHR object
    return deferred.promise();
  };

  // Override Model.prototype.sync and try to clear cache items if it looks
  // like they are being updated.
  Backbone.Model.prototype.sync = function(method, model, options) {
    // Only empty the cache if we're doing a create, update, patch or delete.
    if (method === 'read') {
      return superMethods.modelSync.apply(this, arguments);
    }

    var collection = model.collection,
        keys = [],
        i, len;

    // Build up a list of keys to delete from the cache, starting with this
    keys.push(Backbone.fetchCache.getCacheKey(model));

    // If this model has a collection, also try to delete the cache for that
    if (!!collection) {
      keys.push(Backbone.fetchCache.getCacheKey(collection));
    }

    // Empty cache for all found keys
    for (i = 0, len = keys.length; i < len; i++) { clearItem(keys[i]); }

    return superMethods.modelSync.apply(this, arguments);
  };

  Backbone.Collection.prototype.fetch = function(opts) {
    opts = _.defaults(opts || {}, { parse: true });
    var key = Backbone.fetchCache.getCacheKey(this, opts),
        data = Backbone.fetchCache._cache[key],
        expired = false,
        attributes = false,
        deferred = new $.Deferred(),
        self = this;

    function setData() {
      self[opts.reset ? 'reset' : 'set'](attributes, opts);
      if (_.isFunction(opts.prefillSuccess)) { opts.prefillSuccess(self); }

      // Trigger sync events
      self.trigger('cachesync', self, attributes, opts);
      self.trigger('sync', self, attributes, opts);

      // Notify progress if we're still waiting for an AJAX call to happen...
      if (opts.prefill) { deferred.notify(self); }
      // ...finish and return if we're not
      else {
        if (_.isFunction(opts.success)) { opts.success(self, attributes, opts); }
        deferred.resolve(self);
      }
    }

    if (data) {
      expired = data.expires;
      expired = expired && data.expires < (new Date()).getTime();
      attributes = data.value;
    }

    if (!expired && (opts.cache || opts.prefill) && attributes) {
      // Ensure that cache resolution adhers to async option, defaults to true.
      if (opts.async == null) { opts.async = true; }

      if (opts.async) {
        nextTick(setData);
      } else {
        setData();
      }

      if (!opts.prefill) {
        return deferred.promise();
      }
    }

    // Delegate to the actual fetch method and store the attributes in the cache
    superMethods.collectionFetch.apply(this, arguments)
      // resolve the returned promise when the AJAX call completes
      .done( _.bind(deferred.resolve, this, this) )
      // Set the new data in the cache
      .done( _.bind(Backbone.fetchCache.setCache, null, this, opts) )
      // Reject the promise on fail
      .fail( _.bind(deferred.reject, this, this) );

    // return a promise which provides the same methods as a jqXHR object
    return deferred.promise();
  };

  // Prime the cache from localStorage on initialization
  getLocalStorage();

  // Exports

  Backbone.fetchCache._superMethods = superMethods;
  Backbone.fetchCache.setCache = setCache;
  Backbone.fetchCache.getCacheKey = getCacheKey;
  Backbone.fetchCache.clearItem = clearItem;
  Backbone.fetchCache.setLocalStorage = setLocalStorage;
  Backbone.fetchCache.getLocalStorage = getLocalStorage;

  return Backbone;
}));
